cpp/src/search/contains_table.cu

```diff
@@ -0,0 +1,202 @@
+/*
+ * Copyright (c) 2022, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <join/join_common_utils.cuh>
+#include <join/join_common_utils.hpp>
+
+#include <cudf/detail/join.hpp>
+#include <cudf/detail/null_mask.hpp>
+#include <cudf/table/experimental/row_operators.cuh>
+#include <cudf/table/table_view.hpp>
+#include <cudf/types.hpp>
+
+#include <rmm/cuda_stream_view.hpp>
+#include <rmm/device_uvector.hpp>
+
+#include <thrust/iterator/counting_iterator.h>
+
+#include <cuco/static_multimap.cuh>
+
+namespace cudf::detail {
+
+namespace {
+
+using cudf::experimental::row::lhs_index_type;
+using cudf::experimental::row::rhs_index_type;
+
+/**
+ * @brief Device functor to create a pair of {hash_value, row_index} for a given row.
+ * @tparam T Type of row index, must be `size_type` or a strong index type.
+ * @tparam Hasher The type of internal hasher to compute row hash.
+ */
+template <typename T, typename Hasher>
+struct make_pair_fn {
+  Hasher const hasher;
+  hash_value_type const empty_key_sentinel;
+
+  __device__ inline auto operator()(size_type const i) const
+  {
+    auto const hash_value = remap_sentinel_hash(hasher(i), empty_key_sentinel);
+    return cuco::make_pair(hash_value, T{i});
+  }
+};
+
+/**
+ * @brief The functor to compare two rows using row hashes and row indices.
+ *
+ * @tparam Comparator The row comparator type to perform row equality comparison from row indices.
+ */
+template <typename Comparator>
+struct pair_comparator_fn {
+  Comparator const d_eqcomp;
+
+  pair_comparator_fn(Comparator const d_eqcomp) : d_eqcomp{d_eqcomp} {}
+
+  template <typename LHSPair, typename RHSPair>
+  __device__ inline bool operator()(LHSPair const& lhs_hash_and_index,
+                                    RHSPair const& rhs_hash_and_index) const
+  {
+    auto const& [lhs_hash, lhs_index] = lhs_hash_and_index;
+    auto const& [rhs_hash, rhs_index] = rhs_hash_and_index;
+    return lhs_hash == rhs_hash ? d_eqcomp(lhs_index, rhs_index) : false;
+  }
+};
+
+/**
+ * @brief The functor to accumulate all nullable columns at all nested levels from a given column.
+ *
+ * This is to avoid expensive materializing the bitmask into a real column when calling to
+ * `structs::detail::flatten_nested_columns`.
+ */
+struct accumulate_nullable_columns {
```

Comment:
This can just be a free function that calls a helper function `accumulate`. No need for the extra machinery of a struct owning a result and releasing.